### PR TITLE
Use predis

### PR DESCRIPTION
**Pull request by lnahiro**
Remove redis component and use predis directly

=> Rebase this branch : https://github.com/M6Web/RedisBundle/pull/24
=> Redis is now RedisClient and extends PredisClient to dispatch event when command is executed
=> Remove deprecated cache adapters
=> Remove coke and use phpcs instead
=> Use https://github.com/b1rdex/predis-compressible for compression
=> Add type hinting
=> remove M6/Bundle/M6Web folders

Todo: 
- [x] Reconnection
- [x] Test it on project
- [x] Update README

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,11 +2,11 @@ language: php
 
 php:
   - 7.1
-  - 7.0
```
Comment (Meroje) [inline .travis.yml]: tag should be a major (v4)

File: README.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
 # RedisBundle [![Build Status](https://travis-ci.org/M6Web/RedisBundle.png?branch=master)](https://travis-ci.org/M6Web/RedisBundle)
 
-symfony2 Bundle on top of m6web/redis-component
+symfony2 Bundle on top of predis
 
-see [m6web/redis-component](https://github.com/M6Web/Redis)
+see [predis/predis]()
```
Comment (Oliboy50) [inline README.md]: empty link?

File: composer.json [hunk 3]
```diff
@@ -1,31 +1,31 @@
...
     ],
+    "config": {
+        "vendor-dir": "vendor",
+        "bin-dir": "bin"
+    },
     "require": {
-        "php": ">=7.0",
-        "m6web/redis-component": "^3.3.0",
-        "doctrine/cache":        "~1.3"
+        "php": ">=7.1",
```
Comment (Oliboy50) [inline composer.json]: major release is coming

**Comment by omansour** [comment]
@lnahiro "Remove timeout/readwritetimeout options from client configuration" I am confused about this ? Why removing these options.  I found them usefulls

File: .atoum.php [hunk 4]
```diff
@@ -1,3 +1,3 @@
 <?php
 
-$runner->addTestsFromDirectory(__DIR__.'/src/M6Web/Bundle/RedisBundle/Tests');
\ No newline at end of file
+$runner->addTestsFromDirectory(__DIR__.'/src/Tests');
```
Comment (valentin-claras) [inline .atoum.php]: New line ;)

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,4 @@
 vendor
-composer.lock
\ No newline at end of file
+composer.lock
+.php_cs.cache
+bin/
```
Comment (valentin-claras) [inline .gitignore]: New line ;)

File: README.md [hunk 6]
```diff
@@ -19,77 +19,75 @@ see [m6web/redis-component](https://github.com/M6Web/Redis)
...
 
-```$this->get('m6redis')``` send the default client. ```this->get('m6redis.sharded')``` the sharded one.
+```$this->get('m6web_redis')``` send the default client. ```this->get('m6web_redis.sharded')``` the sharded one.
 
-for a direct access to the predis object (without consistant hashing) (```servers``` section remains the same) :
+### sf2 events
+
+### list of options in servers configuration
+
+TODO
```
Comment (valentin-claras) [inline README.md]: Is it done ?

File: README.md [hunk 7]
```diff
@@ -98,10 +96,12 @@ m6_redis:
 
 The event ```M6Web\Bundle\RedisBundle\EventDispatcher``` is automaticly dispatched to the redis component. Events are fired with the ```redis.command``` label.
 
+TODO : explain how to change by client
```
Comment (valentin-claras) [inline README.md]: What prevents explaining it now ?

File: src/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -37,8 +39,15 @@ private function addServersSection($rootNode)
...
-                            ->scalarNode('ip')->defaultValue('localhost')->end()
+                            ->scalarNode('host')->defaultValue('localhost')->end()
                             ->scalarNode('port')->defaultValue(6379)->end()
+                            ->integerNode('database')->defaultValue(0)->end()
+                            ->scalarNode('scheme')->defaultValue('tcp')->end()
+                            ->booleanNode('async_connect')->defaultValue('false')->end()
+                            ->booleanNode('persistent')->defaultValue('false')->end()
+                            ->floatNode('timeout')->defaultValue(10)->min(0)->end()
+                            ->floatNode('read_write_timeout')->min(0)->end()
+                            ->integerNode('reconnect')->defaultValue(0)->min(0)->max(5)
```
Comment (valentin-claras) [inline src/DependencyInjection/Configuration.php]: Why `max: 5`, is there a major problem with an higher value?

File: src/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -37,8 +39,15 @@ private function addServersSection($rootNode)
...
                         ->children()
-                            ->scalarNode('ip')->defaultValue('localhost')->end()
+                            ->scalarNode('host')->defaultValue('localhost')->end()
                             ->scalarNode('port')->defaultValue(6379)->end()
+                            ->integerNode('database')->defaultValue(0)->end()
+                            ->scalarNode('scheme')->defaultValue('tcp')->end()
+                            ->booleanNode('async_connect')->defaultValue('false')->end()
+                            ->booleanNode('persistent')->defaultValue('false')->end()
+                            ->floatNode('timeout')->defaultValue(10)->min(0)->end()
+                            ->floatNode('read_write_timeout')->min(0)->end()
```
Comment (valentin-claras) [inline src/DependencyInjection/Configuration.php]: :warning: The documentation mention `readwritetimeout` in the client section not `read_write_timeout` in the server section, which and where is it?
Same goes for the other options, some of them are not even mentioned in the doc,
 maybe we should?

File: src/DependencyInjection/M6WebRedisExtension.php [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+class M6WebRedisExtension extends Extension
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = $this->getConfiguration($configs, $container);
+        $config = $this->processConfiguration($configuration, $configs);
+        $clients = isset($config['clients']) ? $config['clients'] : [];
```
Comment (valentin-claras) [inline src/DependencyInjection/M6WebRedisExtension.php]: `$config['clients'] ?? []`

File: src/Profile/CompressionProfile.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            'SETEX' => 'B1rdex\PredisCompressible\Command\StringSetExpire',
+            'SETNX' => 'B1rdex\PredisCompressible\Command\StringSetPreserve',
+            'GET' => 'B1rdex\PredisCompressible\Command\StringGet',
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getVersion()
```
Comment (valentin-claras) [inline src/Profile/CompressionProfile.php]: static ?

File: src/Profile/CompressionProfile.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        $command = parent::createCommand($commandID, $arguments);
+
+        $compressor = new GzipCompressor(self::DEFAULT_THRESHOLD);
+        $compressProcessor = new CompressProcessor($compressor);
+        $compressProcessor->process($command);
+
+        return $command;
+    }
+
+    public function getSupportedCommands()
```
Comment (valentin-claras) [inline src/Profile/CompressionProfile.php]: static ?

File: src/Redis/RedisClient.php [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+     * @var string
+     */
+    protected $eventClass = null;
+
+    /**
+     * eventName to be dispatched
+     *
+     * @var array
+     */
+    protected $eventName = 'redis.command';
```
Comment (valentin-claras) [inline src/Redis/RedisClient.php]: Use the constant ?

File: src/Redis/RedisSessionHandler.php [hunk 14]
```diff
@@ -65,47 +64,43 @@ public function read($sessionId)
...
     {
         try {
             $ret = $this->redis->del($sessionId);
-            if ($ret > 0) {
-                return true;
-            } else {
-                return false;
-            }
+
+            return $ret > 0 ? true : false;
```
Comment (valentin-claras) [inline src/Redis/RedisSessionHandler.php]: `return $ret > 0` ?

File: src/DependencyInjection/Configuration.php (hunk 9 above)
Comment (lnahiro) [reply src/DependencyInjection/Configuration.php]: i updated from readwritetimeout to read_write_timeout. i will update the doc too

File: src/DependencyInjection/Configuration.php (hunk 8 above)
Comment (lnahiro) [reply src/DependencyInjection/Configuration.php]: Its just the time it takes, nb reconnect * timeout. But i will remove the maximum value

File: src/Profile/CompressionProfile.php (hunk 12 above)
Comment (lnahiro) [reply src/Profile/CompressionProfile.php]: its define in abstract

File: src/Profile/CompressionProfile.php (hunk 11 above)
Comment (lnahiro) [reply src/Profile/CompressionProfile.php]: its define in abstract

**Comment by lnahiro** [comment]
@omansour I restored the timeout/read_time_timeout options in clients configuration

File: src/Tests/Fixtures/client_override_config.yml [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    clients:
+        default:
+            servers:   ["server1"]     # list of servers to use
+            prefix: raoul\             # namespace to use
+            timeout: 5
+            read_write_timeout: 5
+        foo:
+            servers:   ["server1"]     # list of servers to use
+            prefix: raoul\             # namespace to use
+            read_write_timeout: 5
```
Comment (valentin-claras) [inline src/Tests/Fixtures/client_override_config.yml]: Missing new line ;)

**Pull request merged**